Tools/machines/frontier-olcf/submit.sh

```diff
@@ -6,12 +6,155 @@
+# https://docs.olcf.ornl.gov/systems/frontier_user_guide.html#sbcasting-a-binary-with-libraries-stored-on-shared-file-systems
+#SBATCH -C nvme
+
+# In case you are using pywarpx, please uncomment the following line.
+# This is needed to select the correct strategy to broadcast the
+# warpx executable and the associated libraries to all the nodes.
+#is_pywarpx=true
+
+if [ $is_pywarpx ]; then
+
+    # Replace this with the name of the WarpX input script
+    input_script="inputs_3d.py"
+
+    # Select WarpX dimensionality ("1", "2", "3", or "RZ")
+    warpx_dim="3"
+
+else
+    # Replace this with the name of the WarpX executable
+    exe="warpx"
+
+    # Replace this with the name of the WarpX inputfile
+    inputfile="inputs"
+fi
+
+# File to redirect stdandard output to
+output="output.txt"
+
+# It may be a good idea to print the current date
+date
+
+########################################################################
+########## INSTRUCTIONS BELOW SHOULD NOT NORMALLY BE MODIFIED ##########
+########################################################################
+
+# ___ Broadcast WarpX and associated libraries to compute nodes_________
+
+if [ ! $is_pywarpx ]; then
+
+    # From the documentation:
+    # SBCAST executable from Orion to NVMe -- NOTE: ``-C nvme``
+    # is needed in SBATCH headers to use the NVMe drive
+    # NOTE: dlopen'd files will NOT be picked up by sbcast
+    sbcast --send-libs --exclude=NONE -pf ${exe} /mnt/bb/$USER/warpx
+    if [ ! "$?" == "0" ]; then
+        # CHECK EXIT CODE. When SBCAST fails, it may leave partial
+        # files on the compute nodes, and if you continue to launch srun,
+        # your application may pick up partially complete shared library files,
+        # which would give you confusing errors.
+        echo "SBCAST failed!"
+        exit 1
+    fi
+    # SBCAST sends all libraries detected by `ld` (minus any excluded),
+    # and stores them in the same directory in each node's node-local storage
+    # Any libraries opened by `dlopen` are NOT sent, since they are not
+    # known by the linker at run-time.
+    # All required libraries now reside in /mnt/bb/$USER/warpx_libs
+    export LD_LIBRARY_PATH="/mnt/bb/$USER/warpx_libs"
+    # libfabric dlopen's several libraries:
+    export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:$(pkg-config --variable=libdir libfabric)"
+    # cray-mpich dlopen's libhsa-runtime64.so and
+    # libamdhip64.so (non-versioned), so symlink on each node:
+    srun -N ${SLURM_NNODES} -n ${SLURM_NNODES} --ntasks-per-node=1 --label -D /mnt/bb/$USER/warpx_libs \
+        bash -c "if [ -f libhsa-runtime64.so.1 ]; then ln -s libhsa-runtime64.so.1 libhsa-runtime64.so; fi;
+        if [ -f libamdhip64.so.5 ]; then ln -s libamdhip64.so.5 libamdhip64.so; fi"
+    # You may notice that some libraries are still linked
+    # from /sw/frontier, even after SBCASTing.
+    # This is because the Spack-build modules use RPATH to find their dependencies.
```

Comment:
Comments/details/notes:
RPATH indeed precedes LD_LIBRARY_PATH. But, these days compilers default to RUNPATH, with is preceded by LD_LIBRARY_PATH at load time.

Spack can be configured to one or the other and defaults currently to RPATH:
https://spack.readthedocs.io/en/latest/config_yaml.html#shared-linking-type

Generally, RPATH here is not too bad because it will not search widely on the filesystem for the dependency. At the same time, we must hope that they use an optimized software FS for their modules, which I assume they do.